### PR TITLE
Autofuzz: JVM branch

### PR DESCRIPTION
**Pull request by arthurscchan**
Generalize some of the function for multiple language of the auto fuzzer generation tool.
Add in JVM branch for the auto fuzzer generation tool.

### CODE REVIEW

File: tools/auto-fuzz/base_files.py [hunk 1]
```diff
@@ -27,8 +26,72 @@ def gen_dockerfile(github_url, project_name):
...
+def gen_project_yaml(github_url, language="python"):
+    BASE_YAML = """fuzzing_engines:
+- libfuzzer
+homepage: %s
+language: %s
+main_repo: %s
+sanitizers:
+- address
+- undefined
+primary_contants: autofuzz@fuzz-introspector.com""" % (github_url, language,
```
Comment (DavidKorczynski) [inline tools/auto-fuzz/base_files.py:87]: nit: `primary_contacts`

File: tools/auto-fuzz/fuzz_driver_generation_jvm.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+        return "data.consumeFloat()"
+    elif argType == "char" or argType == "java.lang.Character":
+        return "data.consumeCharacter()"
+    elif argType == "java.lang.String":
+        return "data.consumeString(100)"
+    else:
+        return _handle_object_creation(argType, init_dict, possible_target)
+
+
+def _search_conrete_subclass(classname, init_dict):
```
Comment (DavidKorczynski) [inline tools/auto-fuzz/fuzz_driver_generation_jvm.py:155]: nit, should be: `_search_concrete_subclass`

File: tools/auto-fuzz/manager.py [hunk 3]
```diff
@@ -19,35 +19,46 @@
...
+import fuzz_driver_generation_jvm
 
 from typing import List, Any
 from multiprocessing.dummy import Pool as ThreadPool
 
 # Set default directories and error if they do not exist
+tqdm_tracker = None
 basedir = os.path.dirname(os.path.realpath(__file__))
-FUZZ_INTRO_BASE = basedir + "/../../"
+FUZZ_INTRO_BASE = basedir + "/../../../fuzz-introspector"
```
Comment (DavidKorczynski) [inline tools/auto-fuzz/manager.py:47]: I don't think we should change this? considering the current location of `auto-fuzz` in `tools/auto-fuzz`

**Review by DavidKorczynski — CHANGES_REQUESTED** [review]
Minor changes. Overall looks great.

File: tools/auto-fuzz/manager.py (hunk 3 above)
Comment (arthurscchan) [reply tools/auto-fuzz/manager.py]: Changed.

File: tools/auto-fuzz/fuzz_driver_generation_jvm.py (hunk 2 above)
Comment (arthurscchan) [reply tools/auto-fuzz/fuzz_driver_generation_jvm.py]: changed.

File: tools/auto-fuzz/base_files.py (hunk 1 above)
Comment (arthurscchan) [reply tools/auto-fuzz/base_files.py]: changed.

File: tools/auto-fuzz/fuzz_driver_generation_jvm.py [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+
+def _generate_heuristic_1(yaml_dict, possible_targets):
+    """Heuristic 1.
+    Creates a FuzzTarget for each method that satisfy all:
+        - public class method which are not abstract or found in JDK library
+        - have between 0-20 arguments
+        - do not have "test" in the function name
+    The fuzz target is simply one that calls into the target function with
+    a string seeded with fuzz data.
```
Comment (DavidKorczynski) [inline tools/auto-fuzz/fuzz_driver_generation_jvm.py:218]: This heuristic generates more than strings as far as I can tell?

File: tools/auto-fuzz/fuzz_driver_generation_jvm.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+            continue
+        if len(func_elem['argTypes']) > 20:
+            continue
+        if "test" in func_elem['functionName']:
+            continue
+
+        possible_target = FuzzTarget()
+
+        # Store target method name
+        func_name = func_elem['functionName'].split('].')[1].split('(')[0]
```
Comment (DavidKorczynski) [inline tools/auto-fuzz/fuzz_driver_generation_jvm.py:244]: Could we add a simple example  as a comment on why this string manipulation is necessary?

**Pull request merged**